File: frontend/src/lib/local/rows.ts
// The rows the per-user stores hold (PLAN.md § On the device → IndexedDB
// stores). Base, overlay and staging rows share one shape per table: the
// server's wire row, with sequences as numbers (they stay far below 2^53),
// `shuffle_seed` and `questions_hash` as unsigned decimal text as on the wire
// (PQ-012), and `seq`, the sequence of the response that wrote the row
// (§ The sync cycle → Pull, 3).
import type { Grade, Origin, Outcome, Progression } from '$lib/cascade/rules';

export type QuizType = 'anagram' | 'definition' | 'leave_value';
export type QuizStatus = 'active' | 'cleared';

export interface CascadeRow {
	id: string;
	name: string;
	quiz_type: QuizType;
	lexicon: string;
	letter_distribution: string;
	clear_threshold: number;
	segment_size: number;
	progression: Progression;
	require_alphabetical: boolean;
	options_changed_at: string;
	options_seq: number;
	options_device_id: string;
	question_count: number;
	depth: number;
	peak_depth: number;
	attempts_since_completion: number;
	created_at: string;
	last_activity_at: string;
	completed_at: string | null;
	trashed_at: string | null;
	updated_seq: number;
	seq: number;
	/** Overlay only: the outbox purged it. */
	deleted?: boolean;
}

export interface QuizRow {
	id: string;
	cascade_id: string;
	level: number;
	origin: Origin;
	origin_quiz_id: string | null;
	origin_attempt: number | null;
	origin_segment_end: number | null;
	status: QuizStatus;
	segment_chain: boolean;
	segment_size: number;
	progression: Progression;
	require_alphabetical: boolean;
	options_changed_at: string;
	options_seq: number;
	options_device_id: string;
	attempt: number;
	shuffle_seed: string;
	questions_hash: string;
	question_count: number;
	correct_count: number;
	missed_count: number;
	cursor: number;
	cursor_moved_at: string | null;
	cursor_device_id: string | null;
	run_start: number;
	created_at: string;
	created_seq: number;
	last_activity_at: string;
	cleared_at: string | null;
	updated_seq: number;
	seq: number;
	/** Question rows yet to be fetched (§ On the device); a restore with no local rows sets it in the overlay. */
	pending?: boolean;
	/** Overlay only: the outbox purged it. */
	deleted?: boolean;
}

export interface QuestionRow {
	quiz_id: string;
	question_idx: number;
	cascade_id: string;
	/** Computed on the device from the quiz's seed; null until materialised. */
	position: number | null;
	grade: Grade | null;
	graded_at: string | null;
	seq: number;
}

export interface AttemptRow {
	quiz_id: string;
	attempt: number;
	cascade_id: string;
	question_count: number;
	correct_count: number;
	missed_count: number;
	outcome: Outcome;
	shuffle_seed: string;
	finished_at: string;
	updated_seq: number;
	seq: number;
}

/** A `questions` store entry: the key the card shows. */
export interface QuestionKey {
	cascade_id: string;
	idx: number;
	key: string;
}

/** A `cards` store entry: the answer, and which extras it holds. */
export interface CardRow {
	cascade_id: string;
	idx: number;
	answer: unknown;
	definitions: boolean;
	hooks: boolean;
}

export interface Binding {
	action: 'show_next' | 'toggle_grade' | 'previous';
	kind: 'key' | 'mouse_button' | 'wheel';
	code: string;
	ctrl: boolean;
	shift: boolean;
	alt: boolean;
	meta: boolean;
}

export interface PreferencesRow {
	default_clear_threshold: number;
	leave_value_decimals: number;
	anagram_show_definitions: boolean;
	anagram_show_hooks: boolean;
	anagram_answer_mode: 'flashcard' | 'typed';
	default_segment_size: number;
	default_progression: Progression;
	default_require_alphabetical: boolean;
	changed_at: string;
	changed_by_device_id: string | null;
	bindings_changed_at: string;
	bindings_device_id: string | null;
	updated_seq: number;
	bindings: Binding[];
}

export interface DistributionRow {
	name: string;
	tiles: { letter: string; blank_letter: string; value: number; is_vowel: boolean }[];
}

/** An outbox entry: the operation as sent, and what the device needs to find it. */
export interface OutboxEntry {
	device_seq: number;
	/** The cascade it touches; absent for preferences and bindings. */
	cascade_id?: string;
	/** Set on a `move_cursor` only, so a newer one finds the pending one. */
	cursor_quiz?: string;
	op: WireOp;
}

/** An operation in its wire form (PLAN.md § API → POST /api/sync). */
export interface WireOp {
	id: string;
	device_seq: number;
	seen_seq: number;
	at: string;
	type: string;
	[field: string]: unknown;
}

// ---------------------------------------------------------------------------
// From the wire
// ---------------------------------------------------------------------------

type Wire = Record<string, unknown>;

const num = (v: unknown) => (typeof v === 'string' ? Number(v) : (v as number));

/** A pulled cascade row, stamped with the response's sequence. */
export function cascadeFromWire(w: Wire, seq: number): CascadeRow {
	return { ...(w as unknown as CascadeRow), options_seq: num(w.options_seq), updated_seq: num(w.updated_seq), seq };
}

export function quizFromWire(w: Wire, seq: number): QuizRow {
	return {
		...(w as unknown as QuizRow),
		options_seq: num(w.options_seq),
		created_seq: num(w.created_seq),
		updated_seq: num(w.updated_seq),
		seq
	};
}

export function attemptFromWire(w: Wire, cascadeId: string, seq: number): AttemptRow {
	return { ...(w as unknown as AttemptRow), cascade_id: cascadeId, updated_seq: num(w.updated_seq), seq };
}

export function preferencesFromWire(w: Wire): PreferencesRow {
	const bindings = ((w.bindings as Wire[]) ?? []).map((b) => ({
		action: b.action,
		kind: b.kind,
		code: b.code,
		ctrl: b.ctrl,
		shift: b.shift,
		alt: b.alt,
		meta: b.meta
	})) as Binding[];
	return { ...(w as unknown as PreferencesRow), updated_seq: num(w.updated_seq), bindings };
}

File: backend/src/cascade/rows.rs
//! Cascade and quiz rows on the wire (PLAN.md § API → Cascades and sync).
//!
//! Every 64-bit value is the decimal text of the unsigned value, never a JSON
//! number: seeds and hashes are stored as two's-complement `i64` and converted
//! back on the way out. Sequences travel as decimal text too.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::PgConnection;
use uuid::Uuid;

use super::order::from_i64;
use crate::search::wire::QuizType;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[serde(rename_all = "snake_case")]
#[sqlx(type_name = "quiz_progression", rename_all = "snake_case")]
pub enum Progression {
    Ladder,
    Drill,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[serde(rename_all = "snake_case")]
#[sqlx(type_name = "quiz_origin", rename_all = "snake_case")]
pub enum Origin {
    Source,
    ClearReplacement,
    DrillReplacement,
    Descent,
    Segment,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[serde(rename_all = "snake_case")]
#[sqlx(type_name = "quiz_status", rename_all = "snake_case")]
pub enum QuizStatus {
    Active,
    Cleared,
}

/// Serialises a u64 stored as i64 as its unsigned decimal text.
pub mod u64_text {
    use serde::Serializer;

    pub fn serialize<S: Serializer>(v: &i64, s: S) -> Result<S::Ok, S::Error> {
        s.serialize_str(&super::from_i64(*v).to_string())
    }
}

/// Serialises a sequence as decimal text.
pub mod seq_text {
    use serde::Serializer;

    pub fn serialize<S: Serializer>(v: &i64, s: S) -> Result<S::Ok, S::Error> {
        s.serialize_str(&v.to_string())
    }
}

#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
pub struct CascadeRow {
    pub id: Uuid,
    pub name: String,
    pub quiz_type: QuizType,
    pub lexicon: String,
    pub letter_distribution: String,
    pub clear_threshold: i16,
    pub segment_size: i32,
    pub progression: Progression,
    pub require_alphabetical: bool,
    pub options_changed_at: DateTime<Utc>,
    #[serde(with = "seq_text")]
    pub options_seq: i64,
    pub options_device_id: Uuid,
    pub question_count: i32,
    pub depth: i32,
    pub peak_depth: i32,
    pub attempts_since_completion: i32,
    pub created_at: DateTime<Utc>,
    pub last_activity_at: DateTime<Utc>,
    pub completed_at: Option<DateTime<Utc>>,
    pub trashed_at: Option<DateTime<Utc>>,
    #[serde(with = "seq_text")]
    pub updated_seq: i64,
}

#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
pub struct QuizRow {
    pub id: Uuid,
    pub cascade_id: Uuid,
    pub level: i32,
    pub origin: Origin,
    pub origin_quiz_id: Option<Uuid>,
    pub origin_attempt: Option<i32>,
    pub origin_segment_end: Option<i32>,
    pub status: QuizStatus,
    pub segment_chain: bool,
    pub segment_size: i32,
    pub progression: Progression,
    pub require_alphabetical: bool,
    pub options_changed_at: DateTime<Utc>,
    #[serde(with = "seq_text")]
    pub options_seq: i64,
    pub options_device_id: Uuid,
    pub attempt: i32,
    #[serde(with = "u64_text")]
    pub shuffle_seed: i64,
    #[serde(with = "u64_text")]
    pub questions_hash: i64,
    pub question_count: i32,
    pub correct_count: i32,
    pub missed_count: i32,
    pub cursor: i32,
    pub cursor_moved_at: Option<DateTime<Utc>>,
    pub cursor_device_id: Option<Uuid>,
    pub run_start: i32,
    pub created_at: DateTime<Utc>,
    #[serde(with = "seq_text")]
    pub created_seq: i64,
    pub last_activity_at: DateTime<Utc>,
    pub cleared_at: Option<DateTime<Utc>>,
    #[serde(with = "seq_text")]
    pub updated_seq: i64,
}

pub const CASCADE_SELECT: &str = r#"
    SELECT c.id, c.name, c.quiz_type, l.name AS lexicon, d.name AS letter_distribution,
           c.clear_threshold, c.segment_size, c.progression, c.require_alphabetical,
           c.options_changed_at, c.options_seq, c.options_device_id, c.question_count, c.depth,
           c.peak_depth, c.attempts_since_completion, c.created_at, c.last_activity_at,
           c.completed_at, c.trashed_at, c.updated_seq
    FROM cascades c
    JOIN lexicons l ON l.id = c.lexicon_id
    JOIN letter_distributions d ON d.id = l.letter_distribution_id"#;

pub const QUIZ_SELECT: &str = r#"
    SELECT q.id, q.cascade_id, q.level, q.origin, q.origin_quiz_id, q.origin_attempt,
           q.origin_segment_end, q.status, q.segment_chain, q.segment_size, q.progression,
           q.require_alphabetical, q.options_changed_at, q.options_seq, q.options_device_id,
           q.attempt, q.shuffle_seed, q.questions_hash, q.question_count, q.correct_count,
           q.missed_count, q.cursor, q.cursor_moved_at, q.cursor_device_id, q.run_start,
           q.created_at, q.created_seq, q.last_activity_at, q.cleared_at, q.updated_seq
    FROM quizzes q"#;

pub async fn cascade(conn: &mut PgConnection, user_id: Uuid, id: Uuid) -> Result<Option<CascadeRow>, sqlx::Error> {
    sqlx::query_as::<_, CascadeRow>(sqlx::AssertSqlSafe(format!("{CASCADE_SELECT} WHERE c.id = $1 AND c.user_id = $2")))
        .bind(id)
        .bind(user_id)
        .fetch_optional(conn)
        .await
}

pub async fn source_quiz(conn: &mut PgConnection, cascade_id: Uuid) -> Result<Option<QuizRow>, sqlx::Error> {
    sqlx::query_as::<_, QuizRow>(sqlx::AssertSqlSafe(format!(
        "{QUIZ_SELECT} WHERE q.cascade_id = $1 AND q.origin = 'source'"
    )))
    .bind(cascade_id)
    .fetch_optional(conn)
    .await
}

pub fn seed_text(v: i64) -> String {
    from_i64(v).to_string()
}
